### PR TITLE
Zbs275/explore zap improvements

### PR DESCRIPTION
**Pull request by zack-erisyon**
A refactor to make the zap context totally separate from the workers.
This eliminates all the redundant args to those functions.
Most importantly, I added a disallow_inner_parallelism which is the real reason for this... it's goal is to make it so that if you have an outer parallel (eg by field) and an inner parallel (over images) that the default behavior is to disable inner parallel if the outer parallel is set. This will prevent contention in most cases but still allow me to have parallel when I call an inner-funciton from a notebook.
I had to refactor every call, as you review please sanity check that I transferred the args to the zap Context correctly.
Note old argument _process_mode  and _debug_mode replace with "mode" which is now "process" (defulat), "thread", "debug"
Moved some terminal display helpers from utils to log
integration and unit tests passing
Fixed a little issue in integrqtion taht that was causing misalignment

### CODE REVIEW

File: plaster/tools/zap/zap.py [hunk 1]
```diff
@@ -51,31 +120,66 @@
...
-    orig_debug_mode = global_debug_mode
-    orig_progress = global_progress
-    global_cpu_limit = cpu_limit
-    global_debug_mode = debug_mode
-    global_progress = progress
-    yield
-    global_cpu_limit = orig_cpu_limit
-    global_debug_mode = orig_debug_mode
-    global_progress = orig_progress
+def Context(
```
Comment (danwinkler) [inline plaster/tools/zap/zap.py:139]: An improvement could be to add a docstring here showing the "default" values for each of these args, as the "default" in this context isn't the default argument but rather what the underscore global var is for each one. Having the docstring here would be nice for IDE autocompletion.

File: plaster/tools/zap/zap.py [hunk 2]
```diff
@@ -1,12 +1,80 @@
...
+                dict(fn=_do_work, args=(i,), foo=i*10)
+                for i in range(10)
+            ])
+
+    Context options:
+        * cpu_limit: Integer. Maximum number of CPUs to use
+            None: all
+            positive numbers: that many cpus
+            negative numbers: all cpus except this many. eg: -2 = all cpus less two
+        * mode:
```
Comment (danwinkler) [inline plaster/tools/zap/zap.py:55]: List default (looks to be "process" from the code")

**Review by danwinkler — APPROVED** [review]
Looks good

File: plaster/tools/zap/zap.py (hunk 1 above)
Comment (zack-erisyon) [reply plaster/tools/zap/zap.py:109]: Moved doc string from above to here.
Clarified defaults.

**Pull request merged**